### PR TITLE
feat(data-table): Added row selection feature to data table

### PR DESCRIPTION
**Pull request by abhiomkar**
## Row Selection

data_table_demo

### Usage

Example usage of data table for programmatically changing row selection.

```ts
dataTable.setSelectedRowIds(['u2', 'u3']);
dataTable.getSelectedRowIds(); // ['u2', 'u3']
```

Make sure to call `.layout()` method when rows are added or removed from data table. This'll re-initializes row checkboxes and updates the state of header checkbox accordingly.

### Events

* `ROW_SELECTION_CHANGED<MDCDataTableRowSelectionChangedEventDetail>`: Triggered when a row checkbox is checked / unchecked.

```ts
interface MDCDataTableRowSelectionChangedEventDetail {
  rowIndex: number;
  rowId: string | null;
  selected: boolean;
}
```

* `SELECTED_ALL`: Triggered when header row checkbox is checked. Checking header row checkbox also selects all row checkboxes.
* `UNSELECTED_ALL`: Triggered when header row checkbox is unchecked. Unchecking header row checkbox also unchecks all row checkboxes.

### Unit Tests

Used [lit-html](https://lit-html.polymer-project.org) for templating component fixture in unit tests of `mdc-data-table/component.ts`.

## TODO

* ~~Add README.md~~
* [DONE] Add more unit tests for 100% test coverage.
* ~~Update screenshot pages to use templates based on `lit-html`.~~

Refs #4591

### CODE REVIEW

File: packages/mdc-data-table/foundation.ts [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+   */
+  private setHeaderRowCheckboxState_() {
+    if (this.adapter_.getSelectedRowCount() === this.adapter_.getRowCount()) {
+      this.adapter_.setHeaderRowCheckboxIndeterminate(false);
+      this.adapter_.setHeaderRowCheckboxChecked(true);
+    } else if (this.adapter_.getSelectedRowCount() === 0) {
+      this.adapter_.setHeaderRowCheckboxIndeterminate(false);
+      this.adapter_.setHeaderRowCheckboxChecked(false);
+    } else {
+      this.adapter_.setHeaderRowCheckboxIndeterminate(true);
```
Comment (bwobrien) [inline packages/mdc-data-table/foundation.ts]: Explicit call to `this.adapter_.setHeaderRowCheckboxChecked(false)` here, so we don't bake in the assumption that setting indeterminate to true will handle it.

File: test/unit/mdc-data-table/foundation.test.js [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+test('#layout should set header row checkbox unchecked when all row checkboxes are unchecked', () => {
+  const {foundation, mockAdapter} = setupTest();
+
+  td.when(mockAdapter.getSelectedRowCount()).thenReturn(0);
+  td.when(mockAdapter.getRowCount()).thenReturn(5);
+
+  foundation.layout();
+  td.verify(mockAdapter.setHeaderRowCheckboxIndeterminate(false), {times: 1});
+  td.verify(mockAdapter.setHeaderRowCheckboxChecked(false), {times: 1});
+});
```
Comment (bwobrien) [inline test/unit/mdc-data-table/foundation.test.js]: I think this is missing the indeterminate case?

Comment (abhiomkar) [reply test/unit/mdc-data-table/foundation.test.js]: It was covered in another unit test case. But I added additional unit test case for `#layout` with indeterminate case.

File: packages/mdc-data-table/foundation.ts (hunk 1 above)
Comment (abhiomkar) [reply packages/mdc-data-table/foundation.ts]: Done.

**Pull request merged**